src/files.ts

```diff
@@ -187,19 +187,48 @@ export async function collectIgnoreRules(
     // Check if symlink and exclude if requested
     if (!fileStats || (fileStats.isSymbolicLink() && !symlinksEnabled) || fileStats.isFile()) return [];
 
-    // Find ignore files inside this directory
+    // Parse top-level .snyk file inside this directory
+    const topLevelSnykIgnores = parseFileIgnores(`${folder}/${DOTSNYK_FILENAME}`);
```

Comment:
Yes, while the issues we've had reported all seemed to be related to the root `.snyk` file, any nested `.snyk` files should also be included in the initial exclude rule collection step.
At the same time, running a full recursive `.snyk` file collection might adversely impact durations, and we would still need to respect the excludes from parent directories (i.e. not collect `.snyk` files from subfolders which are marked excluded from higher-level `.snyk` files), which could complicate the logic significantly.
Will investigate, for the sake of correctness.

Reply:
Implemented support for nested .snyk files by splitting the process into 4 main steps:
- global `.snyk` and `.[*]ignore` file collection
- `.snyk` file exclude rule parsing
- `.[*]ignore` file ignore rule parsing
- filtering and merging ignore and exclude rules

The only caveat with this approach is that it does not account for .snyk files that are excluded by other .snyk files. In order to support that we'd need to do 2 separate passes of `fast-glob`, which practically doubles the duration, or otherwise implement some mechanism for dynamically applying ignore rules while recursively collecting files, which no library seems to support.

Considering most 3rd party packages will not contain `.snyk` files, or if they do, they would only contain positive excludes and not negative ones like `.[*]ignore` files can, I don't see any risk of this adversely impacting the behaviour (e.g. unexpectedly overriding higher-level excludes, or excluding paths from higher-level directories they have no business excluding).

cc @shinin-light as it also relates to one of your points ("relying on a single exception").

Reply:
I'll give it a proper review today or tomorrow (probably tomorrow).

>The only caveat with this approach is that it does not account for .snyk files that are excluded by other .snyk files. In order to support that we'd need to do 2 separate passes of fast-glob, which practically doubles the duration, or otherwise implement some mechanism for dynamically applying ignore rules while recursively collecting files, which no library seems to support.

q: Aren't we already doing 2 passes here? One for collecting the ignore files and one for collecting the files? Doing a 1.5th pass just over the ignore files shouldn't be expensive (but see below).

We probably don't need to address the following in this iteration, but we may want to start questioning some fundamental assumptions in this design e.g.:
- Should we even be using fast-glob at all, or should we just have your own implementation for walking the directory tree with filtering? It seems we are doing a lot of work to make the behavior fit fast-glob.
- Is the collect rules/apply rules 2 pass design the right choice or should we just be doing a single pass "collect files" step for performance reasons?

> Considering most 3rd party packages will not contain .snyk files, or if they do, they would only contain positive excludes and not negative ones like .[*]ignore files can, I don't see any risk of this adversely impacting the behaviour (e.g. unexpectedly overriding higher-level excludes, or excluding paths from higher-level directories they have no business excluding).

Do I understand correctly that you are referring to this case (dummy syntax for ease of readability):
```
# /.snyk

exclude:
 - /sub/.snyk
```

```
# /sub/.snyk

exclude:
  - ./foo.txt
```

If I understand correctly, the "expected behavior" you point out is that the exclude from `/.snyk` should invalidate the exclude from `/sub/.snyk`. In your implementation this wouldn't happen and `/sub/foo.txt` would get excluded. 

I think, if that is the only divergence, that is acceptable and  I'm not even sure whether we would even want `.snyk` files to exclude other `.snyk` files.

Reply:
> In order to support that we'd need to do 2 separate passes of fast-glob, which practically doubles the duration, or otherwise implement some mechanism for dynamically applying ignore rules while recursively collecting files

There is a third option that is less fs-heavy and more computationally heavy. Instead of creating a list of "path exclusion" we could create a `map` of "exclusion source file path" to "excluded path". If we can have a correctly ordered list of all the exclusion rules (positive and negatives both added to the same list) we could definitely parse the rules in order and check their source to know if they should apply or not (based on previous rules in the list). Once the synthesis is done, we don't need any additional fast-glob pass.

> Should we even be using fast-glob at all

It's very fast, although a bit inflexible to our desires

> Is the collect rules/apply rules 2 pass design the right choice or should we just be doing a single pass "collect files" step for performance reasons?

It's a 2-passes design like it always was because the first pass is to collect all the ignore rules (`collectIgnoreRules` function calls fast-glob internally) and another pass is to collect all the valid file paths using the computed file exclusions + the extension filters coming from the API. The 1-pass design is theoretically possible, but we would need to re-write the whole file-scraping logic to either:
- add what @novalex is referring above as "dynamically applying ignore rules while recursively collecting files", which fast-glob just doesn't allow us to do (afaik)
- run a single pass by both creating the `map` mentioned above and also collect a list of `all the file paths` and then run in memory all the exclusion rule computations AND the respective application to the paths list. I'm not sure this is performant tho, since `fast-glob` is specifically designed to be very fast (and not memory intensive) when you know what you're looking for (i.e. using filters), while trying to create a full description of the customer's FS in memory might be too slow for big repos (thousands of ignored paths/dependencies). We could still make a compromise here, since most of the interesting exclusion rules are in the root, we could have 2-passes but with the first pass addressing only root files and the second pass computing the path list and exclusion rules map. This approach would be "correct" because you're not creating out-of-order exceptions to the exclusion logic since the root files are by definition the highest priority.

Reply:
For @strassl-snyk 

> q: Aren't we already doing 2 passes here? One for collecting the ignore files and one for collecting the files? Doing a 1.5th pass just over the ignore files shouldn't be expensive (but see below).

Yes, but the first pass is only looking for ignore files, and the second one looks for all supported files but respecting the ignore rules collected in the first stage. If we were to do yet another pass to look for ignore files but only in the directories not excluded by .snyk files, it would still have to do most of the heavy-lifting to walk the entire dir structure, so in the benchmark the duration for `collectIgnoreRules` was practically doubled.

> Should we even be using fast-glob at all, or should we just have your own implementation for walking the directory tree with filtering? It seems we are doing a lot of work to make the behavior fit fast-glob.

There are a lot of libraries for this, all with drawbacks and advantages, but we're not really doing a lot of work to make it work with `fg` in particular, but rather to try to deal with the limitation of "no dynamic ignores" they all have. I'm quite sure trying to reinvent the wheel and cooking up our own solution would be more work than implementing stages and early filtering to construct the rules we provide to `fg`.

> Is the collect rules/apply rules 2 pass design the right choice or should we just be doing a single pass "collect files" step for performance reasons?

If we do a single pass without any ignore rules, the performance can actually be a lot worse, because we're looking at *all* directories in the structure for *all* supported files. Whereas with the 2-pass implementation we're only looking for supported .snyk and ignore files in all the directories, and then only looking for supported files in the directories that have not been ignored (which most times would include `node_modules` for example) so it's able to scan supported files quicker.

> Do I understand correctly that you are referring to this case (dummy syntax for ease of readability)

Yes, I was more thinking the general "exclude `/sub` from root" and in that case "stop looking at any .snyk files in there as well", but now I realise the deduplication step would remove any rules coming from `/sub/.snyk` anyway.
Your example is more salient, but as you said excluding `.snyk` files themselves is probably not something we want to support, and `.snyk` excludes cannot override higher-level exclude rules like negative matches can, so I think it's a non-issue.
Though we could theoretically parse `.snyk` files in order and skip parsing files under paths that were previously excluded by an already parsed file, it would increase complexity and likely still not work 100% correctly (due to file order as returned by fg, for example), so IMO not worth doing.

For @shinin-light 

> There is a third option that is less fs-heavy and more memory/computation intensive. Instead of creating a list of "path exclusion" we could create a map of "exclusion source file path" to "excluded path". If we can have a correctly ordered list of all the exclusion rules (positive and negatives both added to the same list) we could definitely parse the rules in order and check their source to know if they should apply or not (based on previous rules in the map). Once the synthesis is done, we don't need any additional fast-glob pass.

True, but I think guaranteeing file order is not going to be trivial, and also we'd still need to parse all ignore files first. Given a contrived example of a `.snyk` excluded directory having 100 ignore files, we'd still need to loop and parse through all of those 100 files only to collect the ignore rules that will be discarded in the filtering step. THat's why IMO looking at only `.snyk` files first and then the other ignore files not excluded by `.snyk` is about as efficient we can get, without doing more advanced wizardry like dynamic ignores as we loop through files (and that has it's own challenges including the "guaranteeing file order" one).

> run a single pass by both creating the map mentioned above and also collect a list of all the file paths. Then in memory we can perform all the exclusion rule computations AND the respective application to the paths list. I'm not sure this is performant tho, since fast-glob is specifically designed to be very fast (and not memory intensive) when you know what you're looking for (i.e. using filters), while trying to create a full description of the customer's FS in memory might be too slow for big repos (thousands of ignored paths/dependencies).

That has the potential of killing the CLI process with large-enough repos. Even the array of supported files (respecting any exclude and ignore rules) can be large enough to cause problems like OOM (see https://snyksec.atlassian.net/browse/ZEN-796). Collecting ALL files in memory would only make it worse.

Reply:
> Yes, but the first pass is only looking for ignore files, and the second one looks for all supported files but respecting the ignore rules collected in the first stage. If we were to do yet another pass to look for ignore files but only in the directories not excluded by .snyk files, it would still have to do most of the heavy-lifting to walk the entire dir structure, so in the benchmark the duration for collectIgnoreRules was practically doubled.

What I am saying is we could do an in-memory pass just over the ignore files after we have collected them, not a full pass over the filesystem.

> There are a lot of libraries for this, all with drawbacks and advantages, but we're not really doing a lot of work to make it work with fg in particular, but rather to try to deal with the limitation of "no dynamic ignores" they all have. I'm quite sure trying to reinvent the wheel and cooking up our own solution would be more work than implementing stages and early filtering to construct the rules we provide to fg.

> If we do a single pass without any ignore rules, the performance can actually be a lot worse, because we're looking at all directories in the structure for all supported files. Whereas with the 2-pass implementation we're only looking for supported .snyk and ignore files in all the directories, and then only looking for supported files in the directories that have not been ignored (which most times would include node_modules for example) so it's able to scan supported files quicker.

Yes we are. This entire PR (and also the previous problematic behavior) is working around the fact that fast-glob doesn't fully map to our use case. When I say "single pass" I mean what you refer to as "dynamic applying excludes" - we traverse the file system once (in a single pass) and apply the excludes during the traversal. There really is no wizardry around it - it should be relatively straightforward depth-first traversal - but, as you say, it is additional work to implement and maintain.

> Though we could theoretically parse .snyk files in order and skip parsing files under paths that were previously excluded by an already parsed file, it would increase complexity and likely still not work 100% correctly (due to file order as returned by fg, for example), so IMO not worth doing.

Agreed. It's questionable of whether we want to support this edge case anyway (and/or what the correct behavior is).